packages/server-wallet/src/onchain-service/index.ts

```diff
@@ -0,0 +1,228 @@
+      );
+    }
+
+    // If the channel is not registered, do not send transactions
+    if (!this.store.has(channelId)) {
+      throw new Error(`Channel is not registered ${channelId}`);
+    }
+
+    // Create the errors object
+    const errors: {[k: number]: string} = {};
+    for (let attempt = 1; attempt < this.retries; attempt += 1) {
+      try {
+        const response = await this.queue.add(() => this._sendTransaction(tx));
+        return response;
+      } catch (e) {
+        // Store error
+        errors[attempt] = e.message;
+
+        // Only retry if it is a known error
+        const knownErr = KNOWN_ERRORS.find(err => e.message.includes(err));
+        if (!knownErr) {
+          throw new Error(
+            `Transaction failed to send with unknown error on attempt ${attempt}/5: ${JSON.stringify(
+              errors
+            )}`
+          );
+        }
+      }
+    }
+    throw new Error(`Unable to send transaction after 5 retries: ${JSON.stringify(errors)}`);
+  }
+
+  private async _sendTransaction(tx: MinimalTransaction): Promise<providers.TransactionResponse> {
+    // Get the onchain record of wallet nonce
+    const chainNonce = await this.wallet.getTransactionCount();
+
+    // Use higher of chain or memory nonce
+    const nonce = this.memoryNonce > chainNonce ? this.memoryNonce : chainNonce;
+
+    // TODO: should we set gas minimums or preferences?
+
+    // Send the transaction
+    const response = await this.wallet.sendTransaction(tx);
+    if (!response.hash) {
+      throw new Error(NO_TX_HASH);
+    }
+
+    // Update the memory nonce after sent to mempool
+    this.memoryNonce = nonce + 1;
+    return response;
+  }
+
+  // NOTE: could also do this in the `wait` callback in `submitTransaction`
+  // but would need more information about what event to parse from the
+  // receipt via the API
+  private _registerAssetHolderCallbacks(assetHolder: Contract): void {
+    assetHolder.on(
+      'Deposited',
+      (destination: string, amountDeposited: BigNumber, destinationHoldings: BigNumber) => {
+        if (!this.channelWallet) {
+          throw new Error(
+            'Must attach channel wallet before registering channel, this will go away in v1'
+          );
+        }
+
+        // Get the destination record
+        const record = this.store.get(destination);
+
+        // If there is no channel registered at destination, ignore event
+        if (!record) {
+          return;
+        }
+        const updated = {
+          ...record,
+          ['Deposited']: {
+            destination,
+            amountDeposited,
+            destinationHoldings,
+          },
+        };
+        this.store.set(destination, updated);
+        // Call the appropriate callback on the wallet
+        this.channelWallet.updateChannelFunding({
+          channelId: destination,
+          amount: amountDeposited.toString(),
+          token: this.assetHolders.get(assetHolder.address)?.assetId,
+        });
```

Comment:
You may want to de-couple the chain event monitoring from the post-event hook. Perhaps this `OnchainService` has some [observables](https://rxjs-dev.firebaseapp.com/guide/observable) it maintains, and you can bind this hook function to each observable. e.g.,
```
[ ethers event monitors ] x 50
                    | sends event to
  [ global deposited observable]
                    |
[ hook attached to 'Deposit' ]
                    |
[ hook calls `updateChannelFunding` ]
```